src/Language/DifferentialDatalog/OVSDB/Compile.hs

```diff
@@ -102,23 +176,44 @@ mkComplexType isinput tname TableColumn{..} t@ComplexType{..} = do
     check (max > 0) (pos t) $ "max bound must be greater than 0"
     check (max /= 1 || isNothing valueComplexType) (pos t)
           $ "Cannot handle key-value pairs when max bound is 1"
-    key <- mkBaseType keyComplexType
+    key <- mkBaseType tkind keyComplexType
     case (min, max) of
-         (1,1) -> return $ Left $ mkColName columnName <> ":" <+> key
-         (0,1) -> return $ Left $ mkColName columnName <> ": option_t<" <> key <> ">"
-         _     -> Right <$> mkSubtable isinput tname columnName t
-
-mkSubtable :: (MonadError String me) => Bool -> String -> String -> ComplexType -> me Doc
-mkSubtable isinput tname cname t@ComplexType{..} = do
-    keytype <- mkBaseType keyComplexType
-    valtype <- maybe (return []) (\v -> (\vt -> ["value:" <+> vt]) <$> mkBaseType v) valueComplexType
-    let cols = [ "_uuid: uuid"
-               , "key:" <+> keytype] ++ valtype
-    let prefix = if isinput then "input" else empty
-    return $ prefix <+> "relation" <+> pp tname <> "_" <> pp cname <> "(" $$
-             (nest' $ vcat $ punctuate comma cols)                        $$
-             ")"
-
-mkBaseType :: (MonadError String me) => BaseType -> me Doc
-mkBaseType (BaseTypeSimple at)   = mkAtomicType at
-mkBaseType (BaseTypeComplex cbt) = mkAtomicType $ typeBaseType cbt
+         (1,1) -> return key
+         (0,1) -> return $ "option_t<" <> key <> ">"
+         _     -> do
+             case valueComplexType of
+                  Nothing -> return $ "Set<" <> key <> ">"
+                  Just v  -> do vt <- mkBaseType tkind v
+                                return $ "Map<" <> key <> "," <> vt <> ">"
+
+mkBaseType :: (MonadError String me) => TableKind -> BaseType -> me Doc
+mkBaseType _     (BaseTypeSimple at)   = mkAtomicType at
+mkBaseType tkind (BaseTypeComplex cbt) | isJust (refTableBaseType cbt) && tkind == TableOutput
+                                       = return "string"
+mkBaseType tkind (BaseTypeComplex cbt) | isJust (refTableBaseType cbt) && elem tkind [TableDeltaPlus, TableDeltaUpdate]
+                                       = return "uuid_or_string_t"
```

Comment:
did you reserve this identifier?

Reply:
No need, OVSDB does not have user-defined types, so there is no risk of collision here.